RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGeneratorGPU.cu

```diff
@@ -2,13 +2,170 @@
 // Author: Felice Pantaleo, CERN
 //
 
+#include <cstdint>
+#include <cuda_runtime.h>
+
 #include "HeterogeneousCore/CUDAUtilities/interface/cudaCheck.h"
+#include "RecoLocalTracker/SiPixelRecHits/interface/pixelCPEforGPU.h"
 #include "CAHitQuadrupletGeneratorGPU.h"
 #include "GPUCACell.h"
 #include "gpuPixelDoublets.h"
-#include <cstdint>
 
 using HitsOnCPU = siPixelRecHitsHeterogeneousProduct::HitsOnCPU;
+using namespace Eigen;
+
+__global__ void
+KernelFastFitAllHits(GPU::SimpleVector<Quadruplet> * foundNtuplets,
+    siPixelRecHitsHeterogeneousProduct::HitsOnGPU const * hhp,
+    int hits_in_fit,
+    float B,
+    Rfit::helix_fit *results,
+    Rfit::Matrix3xNd *hits,
+    Rfit::Matrix3Nd *hits_cov,
+    Rfit::circle_fit *circle_fit,
+    Vector4d *fast_fit,
+    Rfit::line_fit *line_fit)
+{
+  int helix_start = (blockIdx.x * blockDim.x + threadIdx.x);
+  if (helix_start >= foundNtuplets->size()) {
+    return;
+  }
+
+#ifdef GPU_DEBUG
+  printf("BlockDim.x: %d, BlockIdx.x: %d, threadIdx.x: %d, helix_start: %d, cumulative_size: %d\n",
+      blockDim.x, blockIdx.x, threadIdx.x, helix_start, foundNtuplets->size());
+#endif
+
+  hits[helix_start].resize(3, hits_in_fit);
+  hits_cov[helix_start].resize(3 * hits_in_fit, 3 * hits_in_fit);
+
+  // Prepare data structure
+  for (unsigned int i = 0; i < hits_in_fit; ++i) {
+    auto hit = (*foundNtuplets)[helix_start].hitId[i];
+    //  printf("Hit global_x: %f\n", hhp->xg_d[hit]);
+    float ge[6];
+    hhp->cpeParams->detParams(hhp->detInd_d[hit]).frame.toGlobal(hhp->xerr_d[hit], 0, hhp->yerr_d[hit], ge);
+    //  printf("Error: %d: %f,%f,%f,%f,%f,%f\n",hhp->detInd_d[hit],ge[0],ge[1],ge[2],ge[3],ge[4],ge[5]);
+
+    hits[helix_start].col(i) << hhp->xg_d[hit], hhp->yg_d[hit], hhp->zg_d[hit];
+
+    for (auto j = 0; j < 3; ++j) {
+      for (auto l = 0; l < 3; ++l) {
+        // Index numerology:
+        // i: index of the hits/point (0,..,3)
+        // j: index of space component (x,y,z)
+        // l: index of space components (x,y,z)
+        // ge is always in sync with the index i and is formatted (guess!!) as:
+        // ge[] ==> [xx, xy, xz, yy, yz, zz]
+        // in (j,l) notation, we have:
+        // ge[] ==> [(0,0), (0,1), (0,2), (1,1), (1,2), (2,2)]
+        // So:
+        // (i=0,j=0,l=1) ==> (x_0,y_0)
+        // (i=1,j=2,l=0) ==> (z_1,x_1)
+        auto ge_idx = j + l;
```

Comment:
wouldn't it be simpler to rewrite this as
```c++
auto ge_idx = j + l + (j > 0 and l > 0);
```
?

Reply:
indeed